exp/addons/internal/controllers/clusterresourceset_scope_test.go

```diff
@@ -0,0 +1,187 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"testing"
+
+	. "github.com/onsi/gomega"
+
+	addonsv1 "sigs.k8s.io/cluster-api/exp/addons/api/v1beta1"
+)
+
+func TestReconcileStrategyScopeNeedsApply(t *testing.T) {
+	tests := []struct {
+		name  string
+		scope *reconcileStrategyScope
+		want  bool
+	}{
+		{
+			name: "no ResourceBinding",
+			scope: &reconcileStrategyScope{
+				baseResourceReconcileScope: baseResourceReconcileScope{
+					resourceSetBinding: &addonsv1.ResourceSetBinding{},
+					resourceRef: addonsv1.ResourceRef{
+						Name: "cp",
+						Kind: "ConfigMap",
+					},
+				},
+			},
+			want: true,
+		},
+		{
+			name: "not applied ResourceBinding",
+			scope: &reconcileStrategyScope{
+				baseResourceReconcileScope: baseResourceReconcileScope{
+					resourceSetBinding: &addonsv1.ResourceSetBinding{
+						Resources: []addonsv1.ResourceBinding{
+							{
+								ResourceRef: addonsv1.ResourceRef{
+									Name: "cp",
+									Kind: "ConfigMap",
+								},
+								Applied: false,
+							},
+						},
+					},
+					resourceRef: addonsv1.ResourceRef{
+						Name: "cp",
+						Kind: "ConfigMap",
+					},
+				},
+			},
+			want: true,
+		},
+		{
+			name: "applied ResourceBinding and different hash",
+			scope: &reconcileStrategyScope{
+				baseResourceReconcileScope: baseResourceReconcileScope{
+					resourceSetBinding: &addonsv1.ResourceSetBinding{
+						Resources: []addonsv1.ResourceBinding{
+							{
+								ResourceRef: addonsv1.ResourceRef{
+									Name: "cp",
+									Kind: "ConfigMap",
+								},
+								Applied: true,
+								Hash:    "111",
+							},
+						},
+					},
+					resourceRef: addonsv1.ResourceRef{
+						Name: "cp",
+						Kind: "ConfigMap",
+					},
+					computedHash: "222",
+				},
+			},
+			want: true,
+		},
+		{
+			name: "applied ResourceBinding and same hash",
+			scope: &reconcileStrategyScope{
+				baseResourceReconcileScope: baseResourceReconcileScope{
+					resourceSetBinding: &addonsv1.ResourceSetBinding{
+						Resources: []addonsv1.ResourceBinding{
+							{
+								ResourceRef: addonsv1.ResourceRef{
+									Name: "cp",
+									Kind: "ConfigMap",
+								},
+								Applied: true,
+								Hash:    "111",
+							},
+						},
+					},
+					resourceRef: addonsv1.ResourceRef{
+						Name: "cp",
+						Kind: "ConfigMap",
+					},
+					computedHash: "111",
+				},
+			},
+			want: false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			gs := NewWithT(t)
+			gs.Expect(tt.scope.needsApply()).To(Equal(tt.want))
+		})
+	}
+}
+
+func TestReconcileApplyOnceScopeNeedsApply(t *testing.T) {
```

Comment:
Let's add a test to ensure the code is re-entrant (it completes apply when some of the objects exists)